File: LeetCode/Peng/24SwapNodesInPairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
       if (!head || !head->next) return head;
       ListNode* result = head->next;
       ListNode* last = NULL;
       while (head && head->next) {
           ListNode* temp = head->next;
           head->next = temp->next;
           temp->next = head;
           if (!last) {
               last = head;
           } else {
               last->next = temp;
               last = head;
           }
           head = head->next;
       }
       return result;
    }
}; 

//recursive version
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
       if (!head || !head->next) return head;
       ListNode* temp = head->next;
       head->next = swapPairs(temp->next);
       temp->next = head;
       return temp;
    }
};

File: LeetCode/Hao/24SwapNodesInPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    private void helper(ListNode node) {
        if (node.next == null || node.next.next == null) {
            return;
        }
        
        ListNode tmp = node.next;
        node.next = node.next.next;
        tmp.next = node.next.next;
        node.next.next = tmp;
        helper(tmp);
    }
    
    
    public ListNode swapPairs(ListNode head) {
        ListNode p = new ListNode(0);
        p.next = head;
        
        helper(p);
    
        return p.next;    
    }
}
